channels/channelmapper.go

```diff
@@ -60,79 +113,149 @@ func GetDefaultSyncFunction(scopeName, collectionName string) string {
+	bodyJSON, err := json.Marshal(body)
 	if err != nil {
 		return nil, err
 	}
-	output := result1.(*ChannelMapperOutput)
-	return output, nil
+	docID, _ := body["_id"].(string)
+	revID, _ := body["_rev"].(string)
+	input := ChannelMapperInput{
+		DocID:   docID,
+		RevID:   revID,
+		Body:    string(bodyJSON),
+		OldBody: oldBodyJSON,
+		Meta:    metaMap,
+		UserCtx: userCtx,
+	}
+	return mapper.MapToChannelsAndAccess2(input)
 }
 
-// Javscript max integer value (https://www.ecma-international.org/ecma-262/5.1/#sec-8.5)
-const JavascriptMaxSafeInt = int64(1<<53 - 1)
-const JavascriptMinSafeInt = -JavascriptMaxSafeInt
-
-// ConvertJSONNumbers converts json.Number values to javascript number objects for use in the sync
-// function.  Integers that would lose precision are left as json.Number, as are floats that can't be
-// converted to float64.
-func ConvertJSONNumbers(value interface{}) interface{} {
-	switch value := value.(type) {
-	case json.Number:
-		if asInt, err := value.Int64(); err == nil {
-			if asInt > JavascriptMaxSafeInt || asInt < JavascriptMinSafeInt {
-				// Integer will lose precision when used in javascript - leave as json.Number
-				return value
-			}
-			return asInt
-		} else {
-			numErr, _ := err.(*strconv.NumError)
-			if numErr.Err == strconv.ErrRange {
-				return value
-			}
+// Runs the sync function. Thread-safe.
+func (mapper *ChannelMapper) MapToChannelsAndAccess2(input ChannelMapperInput) (*ChannelMapperOutput, error) {
+	result, err := mapper.service.WithRunner(func(runner js.Runner) (any, error) {
+		ctx := context.Background()
+		if mapper.timeout > 0 {
+			var cancelFn context.CancelFunc
+			ctx, cancelFn = context.WithTimeout(ctx, mapper.timeout)
+			defer cancelFn()
 		}
+		runner.SetContext(ctx)
+		return callSyncFn(runner, input)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return result.(*ChannelMapperOutput), nil
+}
 
-		if asFloat, err := value.Float64(); err == nil {
-			// Can't reliably detect loss of precision in float, due to number of variations in input float format
-			return asFloat
-		}
-		return value
-	case map[string]interface{}:
-		for k, v := range value {
-			value[k] = ConvertJSONNumbers(v)
+// Parsed version of the JSON object returned by the sync-fn wrapper.
+type syncFnResult = struct {
+	RejectionStatus  int
+	RejectionMessage string
+	Channels         []string
+	Access           map[string][]string
+	Roles            map[string][]string
+	Expiry           any
+}
+
+func callSyncFn(runner js.Runner, in ChannelMapperInput) (*ChannelMapperOutput, error) {
+	// Call the sync fn:
+	jsResult, err := runner.Run(in.DocID, in.RevID, in.Deleted, js.JSONString(in.Body), in.OldBody, in.Meta.Key, string(in.Meta.JSONValue), in.UserCtx)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error calling sync fn: %+w", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("unexpected error calling sync fn: %q", err)
```

Reply:
`%w` has the nice side effect of wrapping the original error inside the one being created so it can be examined later.
But using `%+w` seems wrong; the docs don't say it supports the `+` modifier. I'll remove it.